internal/search/exhaustive/service/service.go

```diff
@@ -147,3 +155,89 @@ func (s *Service) ListSearchJobs(ctx context.Context) (jobs []*types.ExhaustiveS
 
 	return s.store.ListExhaustiveSearchJobs(ctx)
 }
+
+func getPrefix(id int64) string {
+	return fmt.Sprintf("%d-", id)
+}
+
+// CopyBlobs copies all the blobs associated with a search job to the given writer.
+func (s *Service) CopyBlobs(ctx context.Context, w io.Writer, id int64) (err error) {
+	ctx, _, endObservation := s.operations.copyBlobs.With(ctx, &err, opAttrs(
+		attribute.Int64("id", id)))
+	defer endObservation(1, observation.Args{})
+
+	// 🚨 SECURITY: only someone with access to the job may copy the blobs
+	_, err = s.GetSearchJob(ctx, id)
+	if err != nil {
+		return err
+	}
+
+	iter, err := s.uploadStore.List(ctx, getPrefix(id))
+	if err != nil {
+		return err
+	}
+
+	return copyBlobs(ctx, iter, s.uploadStore, w)
+}
+
+// discards output from br up until delim is read. If an error is encountered
+// it is returned. Note: often the error is io.EOF
+func discardUntil(br *bufio.Reader, delim byte) error {
+	// This function just wraps ReadSlice which will read until delim. If we
+	// get the error ErrBufferFull we didn't find delim since we need to read
+	// more, so we just try again. For every other error (or nil) we can
+	// return it.
+	for {
+		_, err := br.ReadSlice(delim)
+		if err != bufio.ErrBufferFull {
+			return err
+		}
+	}
+}
+
+func copyBlobs(ctx context.Context, iter *iterator.Iterator[string], uploadStore uploadstore.Store, w io.Writer) error {
+	// keep a single bufio.Reader so we can reuse its buffer.
+	var br bufio.Reader
+	copyBlob := func(key string, skipHeader bool) error {
+		rc, err := uploadStore.Get(ctx, key)
+		if err != nil {
+			_ = rc.Close()
+			return err
+		}
+		defer rc.Close()
+
+		br.Reset(rc)
+
+		// skip header line
+		if skipHeader {
+			err := discardUntil(&br, '\n')
+			if err == io.EOF {
+				// reached end of file before finding the newline. Write
+				// nothing
+				return nil
+			} else if err != nil {
+				return err
+			}
+		}
+
+		_, err = br.WriteTo(w)
+		return err
+	}
+
+	// For the first blob we want the header, for the rest we don't
+	if iter.Next() {
+		key := iter.Current()
+		if err := copyBlob(key, false); err != nil {
+			return err
+		}
+	}
+
+	for iter.Next() {
+		key := iter.Current()
+		if err := copyBlob(key, true); err != nil {
+			return err
+		}
+	}
```

Comment:
accept if you prefer, otherwise ignore me

```suggestion
	// For the first blob we want the header, for the rest we don't
	skipHeader := false
	if iter.Next() {
		key := iter.Current()
		if err := copyBlob(key, skipHeader); err != nil {
			return errors.Wrapf(err, "failed to copy blob %q", key)
		}
		skipHeader = true
	}
```

Reply:
funny, that's very similar to the code [I wrote a couple of commits earlier](https://github.com/sourcegraph/sourcegraph/pull/56593/commits/18f334ca39b6619ee00881e1236a5cdb0e94486b). Happy to accept.